tango_simlib/tango_yaml_tools/base.py

```diff
@@ -0,0 +1,122 @@
+#########################################################################################
+# Copyright 2020 SKA South Africa (http://ska.ac.za/)                                   #
+#                                                                                       #
+# BSD license - see LICENSE.txt for details                                             #
+#########################################################################################
+"""Module that contains the TangoToYAML class that parses a Tango device specification
+   file (xmi, fgo) or a running Tango device into YAML"""
+from pathlib import Path
+
+import yaml
+
+
+class TangoToYAML:
+    """Class that translates a Tango specification file or a running Tango device to
+       YAML."""
+
+    def __init__(self, parser_class):
+        """Initialise TangoToYAML with a parser class
+
+        Parameters
+        ----------
+        parser_class : Python class definition
+            A Python class that implements methods,
+                - `parse`
+                - `get_device_command_metadata`
+                - `get_device_attribute_metadata`
+                - `get_device_properties_metadata`
+            and has the attribute `device_class_name`
+        """
+        self.parser = parser_class()
+
+    def _build_yaml(self):
+        """Build YAML from the parser
+        """
+        data_dict = [
+            {
+                "class": self.parser.device_class_name,
+                "meta": {"commands": [], "attributes": [], "properties": []},
+            }
+        ]
+
+        for command in self.parser.get_device_command_metadata().values():
+            command_data = {
+                "name": command["name"],
+                "dtype_in": command["dtype_in"].name,
+                "dtype_out": command["dtype_out"].name,
+            }
+            for key in ["doc_out", "doc_in", "inherited", "disp_level"]:
+                if key in command:
+                    command_data[key] = command[key]
+            data_dict[0]["meta"]["commands"].append(command_data)
+
+        for attr in self.parser.get_device_attribute_metadata().values():
+            attr_data = {"name": attr["name"]}
+            if "data_type" in attr:
+                attr_data["data_type"] = attr["data_type"].name
+            if "data_format" in attr:
+                attr_data["data_format"] = attr["data_format"].name
+            if "disp_level" in attr:
+                attr_data["disp_level"] = attr["disp_level"].name
+            # pylint insists on this spacing
+            for key in [
```

Comment:
Probably much more descriptive to assign the list as `attr_props` to make it more descriptive

Reply:
Attributes and Properties are different things in devices, so `attr_props` would be confusing

Reply:
Yeah so aside the device's properties, the commands and attributes also have properties (e.g. the identifiers in the list). Takes getting used to but "Tango speaking" people will understand usage of `attr_props`. Still looks neat this way and probably better to leave it as is so any "non-tango" pair of eyes looking at this code dont get confused.